src/rules/avoid-importing-barrel-files.ts

```diff
@@ -0,0 +1,332 @@
+            default: [],
+            uniqueItems: true,
+            items: {
+              type: 'string',
+            },
+          },
+          mainFields: {
+            type: 'array',
+            description: 'Main fields to use to resolve modules',
+            default: [],
+            uniqueItems: true,
+            items: {
+              type: 'string',
+            },
+          },
+          extensions: {
+            type: 'array',
+            description: 'Extensions to use to resolve modules',
+            default: [],
+            uniqueItems: true,
+            items: {
+              type: 'string',
+            },
+          },
+          // schema to match oxc-resolver's TsconfigOptions
+          tsconfig: {
+            type: 'object',
+            description: 'Options to TsconfigOptions',
+            properties: {
+              configFile: {
+                type: 'string',
+                description: 'Relative path to the configuration file',
+              },
+              references: {
+                type: 'array',
+                description: 'Typescript Project References',
+                items: {
+                  type: 'string',
+                },
+              },
+            },
+          },
+          // NapiResolveOptions.alias
+          alias: {
+            type: 'object',
+            description: 'Webpack aliases used in imports or requires',
+          },
+        },
+      },
+    ],
+    messages: {
+      avoidImport:
+        'The imported module "{{specifier}}" is a barrel file, which leads to importing a module graph of {{amount}} modules, which exceeds the maximum allowed size of {{maxModuleGraphSizeAllowed}} modules',
+    },
+  },
+  defaultOptions: [defaultOptions],
+  create(context) {
+    const options = context.options[0] || defaultOptions
+    const maxModuleGraphSizeAllowed = options.maxModuleGraphSizeAllowed
+    const debug = options.debug
+    const amountOfExportsToConsiderModuleAsBarrel =
+      options.amountOfExportsToConsiderModuleAsBarrel
+    const exportConditions = options.exportConditions
+    const mainFields = options.mainFields
+    const extensions = options.extensions
+    const tsconfig = options.tsconfig
+    const alias = options.alias
+
+    const resolutionOptions = {
+      exportConditions,
+      mainFields,
+      extensions,
+      tsconfig,
+      alias,
+    }
+
+    const resolver = new ResolverFactory({
+      tsconfig,
+      alias,
+      conditionNames: exportConditions,
+      mainFields,
+      extensions,
+    })
+
+    return {
+      ImportDeclaration(node) {
+        const moduleSpecifier = node.source.value
+        const currentFileName = context.filename
+
+        if (options?.allowList?.includes(moduleSpecifier)) {
+          return
+        }
+
+        if (node?.importKind === 'type') {
+          return
+        }
+
+        if (builtinModules.includes(moduleSpecifier.replace('node:', ''))) {
```

Comment:
Should use the resolver concept and test with `resolved.path === null` instead.

Reply:
how do we know the difference between a path which doesn't exist and a built in?

if i `sync(currentFile, 'node:path')` i just get a `Cannot find module 'node:path'`. but we need to handle built-ins differently to missing files

Reply:
`null` means built-in.